xen/arch/x86/boot/head.S

```diff
@@ -522,18 +522,25 @@ __start:
         jmp     trampoline_bios_setup
 
 .Lslaunch_proto:
-        /* Save information that TrenchBoot slaunch was used. */
+        /* Upon reaching here, CPU state mostly matches the one setup by the
+         * bootloader with ESP, ESI and EDX being clobbered above. */
+
+        /* Save information that TrenchBoot slaunch was used and the value of
+         * EBP treating it as a parameter for slaunch. */
         movb    $1, sym_esi(slaunch_active)
+        movl    %ebp,sym_esi(slaunch_param)
```

Comment:
> it would require returning two variables (this is possible, SKL does it).

SKL uses magic of SYSV ABI, while stdcall is forced in this case.  We can still return struct though, will just need to read it from memory.

Reply:
https://github.com/TrenchBoot/xen/compare/4de56cbbf531cf9efb2efd4baf58f66db71b5ae9..6c44eb9bcc4f73b6b98e6a6cfd50c0994c85580c#diff-e4b4491b6c4e847098c559822abc0da4b62e431b9339a971a79026a0c3bac48a
and
https://github.com/TrenchBoot/xen/compare/4de56cbbf531cf9efb2efd4baf58f66db71b5ae9..6c44eb9bcc4f73b6b98e6a6cfd50c0994c85580c#diff-584be1f5aaa2f0679e692f6344b2e03670316e2ebdb7f4ba1dc4fbfb513d196c

Used output parameter instead of return value because returning structures seems to be grey area when it comes to ABI, so it's better to not rely on what compiler does.  In assembly the only change is passing address before all other parameters instead of after them, so it's pretty much the same either way.